gdxsv/lbs_discord.go

```diff
@@ -0,0 +1,524 @@
+			if (b.RenpoUsersBuf.Len() + b.ZeonUsersBuf.Len()) > 0 {
+				valueBuf.WriteString("⠀\n") //insert Braille Pattern Blank to create newline at ending
+			}
+			value := valueBuf.String()
+
+			for i, v := range splitEmbedFieldValues(value) {
+				name := "⠀"
+				if i == 0 {
+					name = b.RegionName
+				}
+
+				battleFields = append(battleFields, &discordEmbedField{
+					Name:  name,
+					Value: v,
+				})
+				accumulatedEmbedStringLength += len(name) + len(v)
+			}
+		}
+		for i, fields := range splitEmbedFields(battleFields) {
+			description := ""
+			if i == 0 {
+				description = fmt.Sprintf("💥 **戦闘中 %d 人**", battleUserCount)
+			}
+			payload.Embed = append(payload.Embed, &discordEmbed{
+				Description: description,
+				Color:       13179394,
+				Fields:      fields,
+			})
+			accumulatedEmbedStringLength += len(description)
+		}
+
+		//
+		// Create the json
+		//
+		buffer := &bytes.Buffer{}
+		encoder := json.NewEncoder(buffer)
+		encoder.SetEscapeHTML(false)
+		err := encoder.Encode(payload)
+
+		if err != nil {
+			logger.Error("Failed to create Discord JSON", zap.Error(err))
+			return
+		}
+
+		//embed structure must not exceed 6000 characters
+		if accumulatedEmbedStringLength > 6000 {
+			originalSize := buffer.Len()
+
+			start := time.Now()
+			jsonString := emojiReplacer.Replace(buffer.String())
+			buffer.Reset()
+			buffer.Write([]byte(jsonString))
+			elapsed := time.Since(start)
+
+			logger.Info("Reduce JSON Size!", zap.Int("original size", originalSize), zap.Any("elapsed", elapsed))
+		}
+
+		logger.Info(buffer.String(), zap.Int("length", buffer.Len()))
+
+		//
+		// Send to Discord
+		//
+		send := func(buf *bytes.Buffer) {
+
+			req, err := http.NewRequest("PATCH", conf.DiscordLiveStatusWebhookURL, buf)
+			req.Header.Set("Content-Type", "application/json")
+
+			resp, err := http.DefaultClient.Do(req)
+			if err != nil {
+				logger.Error("Failed to publish to Discord", zap.Error(err))
+				return
+			}
+			defer resp.Body.Close()
+
+			if ratelimitRemaining := resp.Header.Get("x-ratelimit-remaining"); len(ratelimitRemaining) > 0 {
+				discordLiveStatusRateRemaining, _ = strconv.Atoi(ratelimitRemaining)
+			}
+			if ratelimitReset := resp.Header.Get("x-ratelimit-reset"); len(ratelimitReset) > 0 {
+				discordLiveStatusRateResetEpoch, _ = strconv.ParseInt(ratelimitReset, 10, 64)
+			}
+
+			logger.Info("Discord Webhook sent", zap.String("Status", resp.Status))
+			body, err := ioutil.ReadAll(resp.Body)
+			if err != nil {
+				logger.Error("Failed to read response body", zap.Error(err))
+				return
+			}
+			if resp.StatusCode == http.StatusTooManyRequests {
+				logger.Info("Rate limit", zap.Int64("x-ratelimit-reset", discordLiveStatusRateResetEpoch))
+
+				discordLiveStatusRetryTimer = time.AfterFunc(time.Unix(discordLiveStatusRateResetEpoch, 0).Sub(time.Now()), func() {
+					logger.Info("Retrying last request", zap.Int64("epoch", time.Now().Unix()))
+					publish()
```

Comment:
Since the timer runs the function in another goroutine, it is easy to embed bugs and race conditions.
Simply record the time we can request next time and wait for that time in the publish loop.